clients/src/main/java/org/apache/kafka/common/requests/ProduceRequest.java

```diff
@@ -307,6 +308,7 @@ public static byte requiredMagicForVersion(short produceRequestVersion) {
                 return RecordBatch.MAGIC_VALUE_V1;
 
             case 3:
+            case 4:
```

Comment:
@junrao Sure! After this pull request is closed, I will create tickets for all the TODOs mentioned in the discussion of this pull request. I will ping you and Json for review later.